### PR TITLE
Implement split_edges

### LINKED ISSUE #2276: Add TableCollection.split_edges
As discussed in https://github.com/tskit-dev/tskit/pull/2240#issuecomment-1126892173 this is a generally useful function and is a helpful building block for other topology changing operations.

### PR DESCRIPTION
**Pull request by jeromekelleher**
Closes #2276

Taking over part of #2240

### CODE REVIEW

File: c/tskit/trees.c [hunk 1]
```diff
@@ -3253,6 +3253,123 @@ tsk_treeseq_simplify(const tsk_treeseq_t *self, const tsk_id_t *samples,
...
+        if (mutation.edge != TSK_NULL) {
+            mapped_node = split_edge[mutation.edge];
+        }
+        mutation_time = tsk_is_unknown_time(mutation.time) ? node_time[mutation.node]
+                                                           : mutation.time;
+        if (mapped_node != TSK_NULL && mutation_time >= time) {
+            mutation.node = mapped_node;
+        }
+        /* Update the column in-place to save a bit of time. */
+        tables->mutations.node[j] = mutation.node;
```
[the remark this answers is not in the document]
Comment (jeromekelleher) [reply c/tskit/trees.c:3346]: Oh yeah, good spot. Done.

File: c/tskit/trees.c [hunk 2]
```diff
@@ -3253,6 +3253,122 @@ tsk_treeseq_simplify(const tsk_treeseq_t *self, const tsk_id_t *samples,
...
+            u = tsk_node_table_add_row(&tables->nodes, flags, time, population, TSK_NULL,
+                metadata, metadata_length);
+            if (u < 0) {
+                ret = (int) u;
+                goto out;
+            }
+            ret_id = tsk_edge_table_add_row(&tables->edges, edge.left, edge.right, u,
+                edge.child, edge.metadata, edge.metadata_length);
+            if (ret_id < 0) {
+                ret = (int) ret;
```
Comment (benjeffery) [inline c/tskit/trees.c:3317]: Should this be `ret = (int) ret_id;` ?

File: python/_tskitmodule.c [hunk 3]
```diff
@@ -9344,6 +9344,66 @@ TreeSequence_get_genotype_matrix(TreeSequence *self, PyObject *args, PyObject *k
...
+    if (output == NULL) {
+        goto out;
+    }
+    output->tree_sequence = PyMem_Malloc(sizeof(*output->tree_sequence));
+    if (output->tree_sequence == NULL) {
+        PyErr_NoMemory();
+        goto out;
+    }
+    err = tsk_treeseq_split_edges(self->tree_sequence, time, flags,
+        (tsk_id_t) population, metadata, metadata_length, options,
```
Comment (benjeffery) [inline python/_tskitmodule.c:9394]: Not getting why this cast is needed when `population` is already `tsk_id_t`.

File: python/tests/test_lowlevel.py [hunk 4]
```diff
@@ -1535,6 +1535,45 @@ def test_discrete_time(self):
...
+        with pytest.raises(TypeError):
+            f(flags="0")
+        with pytest.raises(TypeError):
+            f(metadata="0")
+        with pytest.raises(TypeError):
+            f(impute_population="0")
+
+    def test_split_edges_bad_population(self):
+        ts = self.get_example_tree_sequence()
+        with pytest.raises(_tskit.LibraryError, match="POPULATION_OUT_OF_BOUNDS"):
```
Comment (benjeffery) [inline python/tests/test_lowlevel.py:1568]: Oh nice! Hadn't realised adding the macros makes testing more explicit.

File: python/tests/test_table_transforms.py [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+    default_population = population is None
+    if not default_population:
+        # -1 is a valid value
+        if population < -1 or population >= ts.num_populations:
+            raise ValueError("Population out of bounds")
+    flags = 0 if flags is None else flags
+    if metadata is None:
+        metadata = tables.nodes.metadata_schema.empty_value
+    metadata = tables.nodes.metadata_schema.validate_and_encode_row(metadata)
+    # This is the easiest way to turn off encoding when calling add_row below
```
Comment (benjeffery) [inline python/tests/test_table_transforms.py:49]: Does this pattern here imply we should have a way to pass in encoded metadata to `add_row`? Maybe `encoded_metadata=`?

**Review by benjeffery — APPROVED** [review]
Wow, nice stuff.
Just a couple of comments:
 - It might be good to test what happens when split is called with `t=unknown_time`?
 - I can't see where the provenance info mentioned in the docstring is recorded?

File: python/tests/test_table_transforms.py (hunk 5 above)
Comment (jeromekelleher) [reply python/tests/test_table_transforms.py:49]: I think so - either that or have a boolean flag for controlling whether we encode or not?

File: c/tskit/trees.c (hunk 2 above)
Comment (jeromekelleher) [reply c/tskit/trees.c:3317]: Ouch, great spot!

File: python/_tskitmodule.c (hunk 3 above)
Comment (jeromekelleher) [reply python/_tskitmodule.c:9394]: Good spot, fixed.

**Comment by jeromekelleher** [comment]
Thanks for the careful review @benjeffery, super helpful.

I think the simplest thing is to error with nonfinite time, and have added tests.

I can't be bothered with adding provenance, it feels like a waste of effort. It's easy enough to add later if we need it.

**Pull request merged**